File: site/bullet.js
var BULLET_SIZE = 3;
var BULLET_LIFE = 75;

class Bullet {
  constructor(point, vector) {
    this.ticks = 0;
    this.dead = false;
    this.center = point;
    this.speed = vector || BULLET_SPEED;
    this.size = BULLET_SIZE;
  }
  
  static fromDict(data) {
    var bullet = new Bullet(
      new Point(data.center.x, data.center.y),
      new Vector(data.speed.x, data.speed.y)
    );
    bullet.setTicks(data.ticks).setDead(data.dead).setSize(data.size);
    return bullet;
  }
  
  setTicks(ticks) {
    this.ticks = ticks;
    return this;
  }
  
  setDead(isDead) {
    this.dead = isDead;
    return this;
  }
  
  setSize(size) {
    this.size = size;
    return this;
  }
  // Repeats ship. Look into JS inheritance.
  move() {
    this.center.x += this.speed.x;
    this.center.y += this.speed.y;
    
    wrapAround(this.center);
  }
  update() {
    this.move();
    this.ticks++;
    if (this.ticks > BULLET_LIFE) {
      this.dead = true;
    }
  }
}

File: websockets/bullet.py
import utils
from mover import Mover
from utils import Point, Vector

class Bullet(Mover):
  def __init__(
      self,
      center = Point(0, 0),
      speed = Vector(0, 0),
      bullet_size = 3,
      lifetime = 75,
      ticks = 0,
      dead = False):
    super().__init__(center, speed)
    self.size = bullet_size
    self.lifetime = lifetime
    self.ticks = ticks
    self.dead = dead
    
  def from_dict(self, data):
    super().from_dict(data)
    
    self.size = data["size"]
    self.lifetime = data["lifetime"]
    self.ticks = data["ticks"]
    self.dead = data["dead"]
    
    # For chaining and list comprehensions
    return self
  
  def to_dict(self):
    new_data = {
      "size": self.size,
      "lifetime": self.lifetime,
      "ticks": self.ticks,
      "dead": self.dead
    }
    oo_data = super().to_dict()
    return {**new_data, **oo_data}
    
  # # Repeats ship. Consider inheritance.
  # def move(self):
  #   self.center.x += self.speed.x
  #   self.center.y += self.speed.y
  #   
  #   self.center = utils.wrap_around(self.center)

  def update(self):
    super().move()
    self.ticks += 1
    if self.ticks > self.lifetime:
      self.dead = True
      
      
      
      
      
      